docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -56,113 +97,145 @@ The current multisig public keys use amino serialization to generate the address
-or managed accounts (ex. different accounts managed by the `group` module).
+The `+` is a bytes concatenation, which doesn't use any separator.
 
-In the `preImage`, the byte value `0` is used as the separator between `prefix` and `contents`. This is a logical
-choice given that `0` is an invalid value for a string character and is commonly used as a null terminator.
+This algorithm is an outcome from a consulting session with a cryptographer.
+Motivation: this algorithm keeps the address relatively small (length of the `typ` doesn't impact on the length of the final address)
+and it's more secure than [post-hash-prefix-proposal] (which uses first 20 bytes of a pubkey hash, that significantly reduce the address space).
+Moreover the cryptographer motivated the choice to add `typ` in the hash to protect against switch table attack.
 
-### Canonical Public Key Address Prefixes
+We use `address.Hash` function for generating address for all accounts represented by a single key:
+* simple public keys: `address.Hash(keyType, pubkey)`
++ aggregated keys (eg: BLS): `address.Hash(keyType, aggregatedPubKey)`
++ module accounts: `addoress.Hash("module", moduleID)`
+  For module subaccount, or module specific content we append a specific key to the moduleKey:
+  `address.Hash("module", moduleSubaccountKey)`  (note: `moduleSubaccountKey` already composes the `modleKey`).
 
-All public key types will have a unique protobuf message type such as:
 
-```proto
-package cosmos.crypto.sr25519;
+### Composed Account Addresses
 
-message PubKey {
-  bytes key = 1;
+For simple composed accounts (like new naive multisig), we generalize the `address.Hash`. The address is constructed by recursively creating addresses for the sub accounts, sorting the addresses and composing it into a single address. It ensures that the ordering of keys doesn't impact the created address.
+
+```go
+// We don't need a PubKey interface - we need anything which is addressable.
+type Addressable interface {
+    Address() []byte
+}
+
+func Composed(typ string, subaccounts []Addressable) []byte {
+    addresses = map(subaccounts, \a -> a.Address())
+    addresses = sort(addresses)
+    return address.Hash(typ, addresses[0] + ... + addresses[n])
 }
 ```
- 
-All protobuf messages have unique fully qualified names, in this example `cosmos.crypto.sr25519.PubKey`.
-These names are derived directly from .proto files in a standardized way and used
-in other places such as the type URL in `Any`s. Since there is an easy and obvious
-way to get this name for every protobuf type, we can use this message name as the
-key type `prefix` when creating addresses. For all basic public keys, `contents`
-should just be the raw unencoded public key bytes.
 
-Thus the canonical address for new public key types would be `AddressHash(proto.MessageName(pk), pk.Bytes)`.
+The `typ` parameter should contain all significant attributes with deterministic serialization.
```

Comment:
This ADR should also define the `typ` for all existing keys in the SDK, right? If the previous iteration, i believe `typ := proto.MessageName(pk)`. If we're keeping that, it would be good to mention it.

Reply:
It's defined in the sections below.